### PR TITLE
[5.0] Generic ModalSelect field

### PR DESCRIPTION
**Pull request by Fedik**
### Summary of Changes

Adding new generic `ModalSelectField` to select content in modal.
This is addittion to #40150 to improve our modal windows.
And should replace (in future) modal-fields.js.
Instead of hardcoded callback, now it should be [used `window.postMessage()`](https://github.com/joomla/joomla-cms/blob/4120d720b642ea61052edab5e65e53b3c19c66c1/build/media_source/system/js/modal-content-select.js#L10-L26) wich is more safe, and all Form buttons should be within same DOM document.

This approcah also will help to improve Editor-xtd buttons #40202.

In this PR I have transfrmed Modal_Article field to use new ModalSelect.
However it still will be need some work to improve handling propagationAssotiation.

### Testing Instructions

Apply #40150 patch and apply current PR.
Go to menu and crate a Single Article menu.
In article field test: Select, Create, Clear buttons.

In multilanguage site try edit Article association.

Also can test the field in subform:
```xml
<field type="subform" name="subform" label="subform" multiple="true">
  <form>
    <field 
        type="ModalSelect"
        name="article" label="Article"
        urlSelect="index.php?option=com_content&amp;view=articles&amp;layout=modal&amp;tmpl=component"
    />
  </form>
</field>
```

### Actual result BEFORE applying this Pull Request
Works

### Expected result AFTER applying this Pull Request
Works but better :)

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: https://docs.joomla.org/Standard_form_field_types
- [ ] No documentation changes for docs.joomla.org needed
- [x] Pull Request link for manual.joomla.org: https://github.com/joomla/Manual/pull/178
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by HLeithner** [comment]
I like the postMessage approach. I'm not sure if it could be a more generic for example to have one "modalreturn" layout for all usages of the modalfield.

**Comment by Fedik** [comment]
>  a more generic for example to have one "modalreturn" layout 

This mainly need to handle "Save & Close" / "Cancel" (in modal) because after save we always have an redirect.

I thinking about generic layout kind of we have for "emptystate". Will see, maybe later will find some better ideas.

**Comment by brianteeman** [comment]
Really excited about this

**Comment by HLeithner** [comment]
> Really excited about this

me too

**Comment by chmst** [comment]
me too!

**Comment by Fedik** [comment]
@brianteeman do you remember what Propagation button actualy doing (in multilanguage)? 
Currently it is coded in to every Article_Modal, Menu_Modal etc. fields. It a bit hacky thing that I have no idea what to do :)

**Comment by brianteeman** [comment]
I have zero idea. Everytime I tried to work it out I failed

**Comment by Fedik** [comment]
Okay, I have found the origin #21321. Still not fully understood, need to think how to keep it.

File: administrator/components/com_content/src/View/Article/HtmlView.php [hunk 1]
```diff
@@ -226,4 +236,41 @@ function (Toolbar $childBar) use ($checkedOut, $itemEditable, $canDo, $user) {
...
+     */
+    protected function addModalToolbar()
+    {
+        $user       = $this->getCurrentUser();
+        $userId     = $user->id;
+        $isNew      = ($this->item->id == 0);
+        $checkedOut = !(is_null($this->item->checked_out) || $this->item->checked_out == $userId);
+        $toolbar    = Toolbar::getInstance();
+
+        // Built the actions for new and existing records.
```
Comment (brianteeman) [inline administrator/components/com_content/src/View/Article/HtmlView.php]:
```suggestion
        // Build the actions for new and existing records.
```

File: libraries/src/Form/Field/ModalSelectField.php [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    /**
+     * Enabled actions: select, clear, edit, new
+     *
+     * @var    boolean[]
+     * @since  __DEPLOY_VERSION__
+     */
+    protected $canDo = [];
+
+    /**
+     * Urls to for modal: select, edit, new
```
Comment (brianteeman) [inline libraries/src/Form/Field/ModalSelectField.php]:
```suggestion
     * Urls for modal: select, edit, new
```

File: administrator/components/com_content/src/Controller/ArticleController.php [hunk 3]
```diff
@@ -56,6 +56,35 @@ public function __construct($config = [], MVCFactoryInterface $factory = null, $
...
+     *
+     * @return  boolean  True if access level checks pass, false otherwise.
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function cancel($key = null)
+    {
+        $result = parent::cancel($key);
+
+        // When is editing in modal then redirect to modalreturn layout
```
Comment (brianteeman) [inline administrator/components/com_content/src/Controller/ArticleController.php]:
```suggestion
        // When editing in modal then redirect to modalreturn layout
```

File: administrator/components/com_content/src/Controller/ArticleController.php [hunk 4]
```diff
@@ -90,6 +119,17 @@ protected function postSaveHook(BaseDatabaseModel $model, $validData = [])
             ]);
 
             $this->setRedirect(Route::_('index.php?option=com_menus&view=item&client_id=0&menutype=mainmenu&layout=edit', false));
+        } elseif ($this->input->get('layout') === 'modal' && $this->task === 'save') {
+            // When is editing in modal then redirect to modalreturn layout
```
Comment (brianteeman) [inline administrator/components/com_content/src/Controller/ArticleController.php]:
```suggestion
            // When editing in modal then redirect to modalreturn layout
```

**Comment by HLeithner** [comment]
I like it, it looks good. only thing I'm not so sure if the buttons in the iframe are a good idea... we don't have much space in the iframe and that's another row...

image

looking at the Ipad air there is not much room for real content
image

**Comment by Fedik** [comment]
> we don't have much space in the iframe and that's another row...

They will eat space wherever you place them, but they should be within the same document where the form is. 

>  there is not much room for real content

I would not expect much room for such editing anyway. 
Default dialog size is:
```
width: 80vw;
max-width: 1700px;
height: 80vh;
```
And can be styled.

**Comment by Fedik** [comment]
I have increased default size

**Comment by Fedik** [comment]
I have prepared changes for Menu field also: https://github.com/Fedik/joomla-cms/compare/popup-field3...popup-field-menu

**Comment by HLeithner** [comment]
thanks, I merge this for beta1 to get better feedback and we have time to fix issues.

**Pull request merged**